Python/ceval.c

```diff
@@ -4986,6 +4987,77 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+            assert(call_shape.kwnames == NULL);
+            _PyCallCache *cache = (_PyCallCache *)next_instr;
+            DEOPT_IF(is_method(stack_pointer, oparg), CALL);
+            PyObject *callable = PEEK(oparg+1);
+            DEOPT_IF(!PyType_Check(callable), CALL);
+            PyHeapTypeObject *tp = (PyHeapTypeObject *)callable;
+            DEOPT_IF(tp->ht_type.tp_version_tag != read_u32(cache->func_version), CALL);
+            STAT_INC(CALL, hit);
+            PyFunctionObject *init = (PyFunctionObject *)tp->_spec_cache.init;
+            PyCodeObject *code = (PyCodeObject *)init->func_code;
+            DEOPT_IF(code->co_argcount != oparg+1, CALL);
+            PyObject *self = _PyType_NewManagedObject(&tp->ht_type);
+            if (self == NULL) {
+                goto error;
+            }
+            PEEK(oparg+1) = self;
+            Py_DECREF(tp);
+            Py_INCREF(_Py_InitCleanupFunc);
+            _PyInterpreterFrame *shim = _PyFrame_Push(tstate, _Py_InitCleanupFunc);
+            if (shim == NULL) {
+                goto error;
+            }
+            CALL_STAT_INC(inlined_py_calls);
+            shim->previous = frame;
+            shim->prev_instr = _PyCode_CODE(shim->f_code) + 1;
+            assert(_Py_OPCODE(*(_PyCode_CODE(shim->f_code)+ 2)) == EXIT_INIT_CHECK);
+            if (_Py_EnterRecursiveCallTstate(tstate, "")) {
+                tstate->recursion_remaining--;
+                goto exit_unwind;
+            }
+            /* Push self onto stack of shim */
+            Py_INCREF(self);
+            shim->stacktop = 1;
+            shim->localsplus[0] = self;
+            Py_INCREF(init);
+            _PyInterpreterFrame *init_frame = _PyFrame_Push(tstate, init);
+            if (init_frame == NULL) {
+                goto error;
+            }
+            /* Copy self followed by args to __init__ frame */
+            STACK_SHRINK(oparg+1);
+            for (int i = 0; i <= oparg; i++) {
+                init_frame->localsplus[i] = stack_pointer[i];
+            }
+            STACK_SHRINK(1); // Clear the NULL.
+            _PyFrame_SetStackPointer(frame, stack_pointer);
+            for (int i = oparg+1; i < code->co_nlocalsplus; i++) {
+                init_frame->localsplus[i] = NULL;
+            }
+            JUMPBY(INLINE_CACHE_ENTRIES_CALL);
+            frame->prev_instr = next_instr - 1;
+            init_frame->previous = shim;
+            frame = cframe.current_frame = init_frame;
+            goto start_frame;
+        }
+
+        TARGET(EXIT_INIT_CHECK) {
+            assert(STACK_LEVEL() == 2);
+            PyObject *should_be_none = TOP();
+            if (should_be_none != Py_None) {
+                PyErr_Format(PyExc_TypeError,
+                    "__init__() should return None, not '%.200s'",
+                    Py_TYPE(should_be_none)->tp_name);
+                goto error;
+            }
+            STACK_SHRINK(1);
+            Py_DECREF(Py_None);
```

Comment:
```suggestion
            _Py_DECREF_NO_DEALLOC(Py_None);
```
This would avoid dealloc branch.